pymc3/tests/test_distributions_random.py

```diff
@@ -402,6 +402,17 @@ class TestCategorical(BaseTestCases.BaseTestCase):
     def get_random_variable(self, shape, with_vector_params=False, **kwargs):  # don't transform categories
         return super(TestCategorical, self).get_random_variable(shape, with_vector_params=False, **kwargs)
 
+    def test_scalar_parameter_shape(self):
+        rv = self.get_random_variable(None)
+        # for size, expected in ((None, (1,)), (5, (5, 1)), ((4, 5), (4, 1))):
+        for size, expected in (((4, 5), (4, 1)),):
```

Comment:
So while I fixed the original inciting case in the issue, I don't know enough about all the possible ways that categorical.random could be called to know how to modify these tests correctly.

The original example passes `p` as a parameter to `dist.random` and that case is handled
This test passes `size` to `dist.random` and I don't understand how that should be handled by `Categorica.random` or if this ever happens during nominal use of this library.

Any guidance would be appreciated and thanks for bearing with me while I learn the ropes.

![image](https://user-images.githubusercontent.com/7213793/42424573-5388a828-82c3-11e8-8ef5-686857121012.png)

Reply:
Just to be clear, in this example, you expected `v.shape` to be `(10, 5)`, right?  I tried to get `draw_values` and `generate_samples` to return something shaped roughly the following way:

1. broadcast the parameters together using `np.broadcast` to get their shape. Call it `d1`.
2. look at the `shape` argument passed to the random variable, and call it `d2`.  Try to combine `d1` and `d2` in a sensible way, and throw an exception instead of guessing _too_ much.  This is the bulk of the `draw_values` code.  We'll say `d3` is the combination of `d1` and `d2`
3. return something of shape `size x d3`

Explicitly, consider `x = pm.Normal.dist(mu=np.random.rand(5), sd=1, shape=(2, 5)).random(size=10)

1. The size of `mu` is `(5,)`, and `sd` is `()`, and broadcasting them gives `(5,)`
2. The shape argument (`(2, 5)`) fits nicely with the broadcast shape, so `d3` above is `(2, 5)`.  Note that `shape=(2, 4)` throws an exception.
3. We return an array of shape `(10, 2, 5)`, since the size argument was 10.

I hope this helps!

Reply:
Hey @ColCarroll 
I believe v.shape should be (10,1) per the issue @junpenglao filed? https://github.com/pymc-devs/pymc3/issues/3035

I've read your comments and want to read them a couple more times before asking any other follow up questions. This is very helpful

Reply:
@ColCarroll and @junpenglao 

Would either of you mind if I put the deduplication of random_shape on another pull request so it can be merged independently of the shape issue? All tests were passing for that portion of the code.

I ask because I think it'll streamline the code changes and reduce the risk of conflicts later.

Reply:
Sure thing. Which PR is that?

Reply:
Just created it here. Pending CI tests at time of writing
https://github.com/pymc-devs/pymc3/pull/3084